--- /home/users/baggins/devel/PLD/rpm/BUILD/libgphoto2-2.5.4/configure.ac~	2014-03-23 18:38:18.000000000 +0100
+++ /home/users/baggins/devel/PLD/rpm/BUILD/libgphoto2-2.5.4/configure.ac	2015-02-26 20:01:07.265114858 +0100
@@ -285,7 +285,7 @@
 dnl ---------------------------------------------------------------------------
 dnl check for libgd
 dnl ---------------------------------------------------------------------------
-GP_CHECK_LIBRARY([LIBGD],[libgd],[>= 2.0],[gd.h],[gdImageCreateTrueColor],[],[],[default-on],[http://www.libgd.org/])
+GP_CHECK_LIBRARY([LIBGD],[gdlib],[>= 2.0],[gd.h],[gdImageCreateTrueColor],[],[],[default-on],[http://www.libgd.org/])
 GP_CONFIG_MSG([libGD conversion support],[${have_LIBGD}])
 
 dnl ---------------------------------------------------------------------------
