--- libgphoto2-2.1.6/camlibs/canon/canon.c.orig	2005-06-07 19:57:47.000000000 +0200
+++ libgphoto2-2.1.6/camlibs/canon/canon.c	2005-06-23 23:14:14.951959648 +0200
@@ -223,7 +223,7 @@
 
         /* 0x3099 is the EOS 300D/Digital Rebel in PTP mode */
         /* A80 seems to share the ID for PTP and Canon modes */
-        {"Canon:PowerShot A80 (normal mode)",   CANON_CLASS_1,  0x04A9, 0x309A, CAP_SUP, SL_MOVIE_LARGE, SL_THUMB, SL_PICTURE, NULL},
+        {"Canon:PowerShot A80 (normal mode)",   CANON_CLASS_5,  0x04A9, 0x309A, CAP_SUP, SL_MOVIE_LARGE, SL_THUMB, SL_PICTURE, NULL},
         /* 0x309b is the SD10 Digital ELPH/Digital IXUS i/IXY Digital L
            in PTP mode; will it work in Canon mode? */
         {"Canon:PowerShot SD10 Digital ELPH (normal mode)",   CANON_CLASS_1,  0x04A9, 0x309B, CAP_SUP, SL_MOVIE_LARGE, SL_THUMB, SL_PICTURE, NULL},
