--- libgphoto2-2.1.4/libgphoto2_port/configure.in.orig	Sun Jan 18 20:47:09 2004
+++ libgphoto2-2.1.4/libgphoto2_port/configure.in	Wed Jul 21 16:02:05 2004
@@ -60,7 +60,7 @@
 dnl ---------------------------------------------------------------------------
 dnl i18n support
 dnl ---------------------------------------------------------------------------
-ALL_LINGUAS="cs da de fr es it ja no pt_BR ru sl sv uk zh_CN"
+ALL_LINGUAS="cs da de fr es it ja nb pt_BR ru sl sv uk zh_CN"
 # EVIL HACKS
 AC_MSG_CHECKING([for po/Makevars requiring evil hack])
 GETTEXT_PACKAGE="${PACKAGE}-${LIBGPHOTO2_PORT_CURRENT}"
--- libgphoto2-2.1.4/configure.in.orig	Sun Jan 18 20:47:32 2004
+++ libgphoto2-2.1.4/configure.in	Wed Jul 21 16:01:50 2004
@@ -50,7 +50,7 @@
 dnl USE_INCLUDED_LIBINTL=no
 dnl AC_SUBST(USE_INCLUDED_LIBINTL)
 
-ALL_LINGUAS="da de es fr hu it ja no ru sl sv uk zh_CN"
+ALL_LINGUAS="da de es fr hu it ja nb ru sl sv uk zh_CN"
 # EVIL HACKS
 AC_MSG_CHECKING([for po/Makevars requiring evil hack])
 GETTEXT_PACKAGE="${PACKAGE}-${LIBGPHOTO2_CURRENT}"
