--- libgphoto2-2.1.1/libgphoto2_port/doc/DocRules.am	Sun Dec  1 21:45:44 2002
+++ libgphoto2-2.1.1.new/libgphoto2_port/doc/DocRules.am	Mon Mar  3 18:30:47 2003
@@ -22,10 +22,10 @@
 templates: scan
 	cd $(srcdir) && gtkdoc-mktmpl --module=$(DOC_MODULE)
 
-sgml:
+sgml: templates
 	cd $(srcdir) && gtkdoc-mkdb --module=$(DOC_MODULE) --source-dir=$(DOC_SOURCE_DIR)
 
-html:
+html: sgml
 	if ! test -d html ; then mkdir html ; fi
 	cd html && gtkdoc-mkhtml $(DOC_MODULE) ../$(srcdir)/$(DOC_MAIN_SGML_FILE)
 
