diff -ur libgphoto2-2.4.1.orig/libgphoto2_port/m4/gp-documentation.m4 libgphoto2-2.4.1/libgphoto2_port/m4/gp-documentation.m4
--- libgphoto2-2.4.1.orig/libgphoto2_port/m4/gp-documentation.m4	2007-07-27 07:53:47.000000000 +0200
+++ libgphoto2-2.4.1/libgphoto2_port/m4/gp-documentation.m4	2008-05-29 15:13:40.000000000 +0200
@@ -107,7 +107,7 @@
 fi
 AC_SUBST([htmldir])
 apidocdir="${htmldir}/api"
-AC_SUBST([apidocdir}])
+AC_SUBST([apidocdir])
 
 ])dnl
 
diff -ur libgphoto2-2.4.1.orig/m4m/gp-documentation.m4 libgphoto2-2.4.1/m4m/gp-documentation.m4
--- libgphoto2-2.4.1.orig/m4m/gp-documentation.m4	2007-07-27 07:54:04.000000000 +0200
+++ libgphoto2-2.4.1/m4m/gp-documentation.m4	2008-05-29 15:13:32.000000000 +0200
@@ -107,7 +107,7 @@
 fi
 AC_SUBST([htmldir])
 apidocdir="${htmldir}/api"
-AC_SUBST([apidocdir}])
+AC_SUBST([apidocdir])
 
 ])dnl
 
